File: server/app/services/pdf_drawer.py
from concurrent.futures import ThreadPoolExecutor
from io import BytesIO
from urllib.parse import urlparse

import requests
from PIL import Image
from reportlab.lib.colors import HexColor, black
from reportlab.lib.pagesizes import A4
from reportlab.lib.units import cm
from reportlab.lib.utils import ImageReader
from reportlab.pdfbase.pdfmetrics import stringWidth
from reportlab.pdfgen import canvas

# ==============================================================================
# CONFIGURATION & LAYOUT CONSTANTS
# ==============================================================================

# --- Page Setup ---
PAGE_W, PAGE_H = A4

MARGIN_L = 1.8 * cm
MARGIN_R = 1.8 * cm
MARGIN_TOP = 1.6 * cm
MARGIN_BOTTOM = 1.8 * cm
CONTENT_W = PAGE_W - MARGIN_L - MARGIN_R

# --- Column Positions & Widths (Detailed Catalog View) ---
COL_NO_X = MARGIN_L
COL_CODE_X = MARGIN_L + 1.0 * cm
COL_DESC_X = MARGIN_L + 3.6 * cm
COL_QTY_X = MARGIN_L + 10.8 * cm
COL_IMG_X = MARGIN_L + 12.6 * cm
COL_IMG_W = CONTENT_W - 12.6 * cm

DESC_COL_W = COL_QTY_X - COL_DESC_X - 0.3 * cm

# --- Typography & Row Heights ---
FONT_BODY, FONT_BODY_SIZE = "Helvetica", 10.5
LINE_HEIGHT = 14

IMG_MAX_W = COL_IMG_W - 0.2 * cm
IMG_MAX_H = 3.6 * cm

ROW_TOP_PAD = 0.55 * cm
ROW_BOTTOM_PAD = 0.55 * cm
MIN_ROW_CONTENT_H = 2.3 * cm

# --- Summary Table Configuration ---
TABLE_HEADER_H = 0.9 * cm
CELL_PAD_X = 0.15 * cm
CELL_PAD_Y = 0.25 * cm
TABLE_BORDER_COLOR = HexColor("#333333")
TABLE_BORDER_WIDTH = 0.6

SUMMARY_COL_BOUNDS = [
    MARGIN_L,
    COL_CODE_X,
    COL_DESC_X,
    COL_QTY_X + 5 * cm,
    PAGE_W - MARGIN_R,
]
SUMMARY_COL_LABELS = ["No", "code", "Description", "Qty"]
SUMMARY_DESC_COL_W = (COL_QTY_X + 5 * cm) - COL_DESC_X - 2 * CELL_PAD_X
SUMMARY_ROW_GAP = 0.35 * cm

# ==============================================================================
# IMAGE FETCHING & PREFETCHING UTILITIES
# ==============================================================================

# Persistent HTTP session for connection pooling and keep-alive optimization
_session = requests.Session()


def _fetch_bytes(url, timeout=6, retries=1):
    """
    Downloads raw bytes for a given URL with a basic retry budget.
    """
    for attempt in range(retries + 1):
        try:
            resp = _session.get(url, timeout=timeout)
            resp.raise_for_status()
            return resp.content
        except Exception as e:
            if attempt == retries:
                print(f"Failed to fetch {url}: {e}")
                return None


def load_image(path_or_url, raw_bytes=None):
    """
    Builds a ReportLab ImageReader instance from:
    - Pre-fetched raw bytes,
    - A local file path, or
    - A remote URL (as fallback if not prefetched).

    Returns None if loading fails.
    """
    if not path_or_url:
        return None
    try:
        if raw_bytes is not None:
            img = Image.open(BytesIO(raw_bytes))
        else:
            parsed = urlparse(path_or_url)
            if parsed.scheme in ("http", "https"):
                data = _fetch_bytes(path_or_url)
                if data is None:
                    return None
                img = Image.open(BytesIO(data))
            else:
                img = Image.open(path_or_url)
        img.load()
    except Exception as e:
        print(f"Failed to load image {path_or_url}: {e}")
        return None

    if img.mode not in ("RGB", "RGBA", "L"):
        img = img.convert("RGB")
    return ImageReader(img)


def prefetch_images(products, max_workers=10):
    """
    Concurrently downloads all remote product images using ThreadPoolExecutor.
    Returns a mapping dict of {url: ImageReader or None}.
    """
    urls = sorted(
        {
            p.get("image")
            for p in products
            if p.get("image") and urlparse(p["image"]).scheme in ("http", "https")
        }
    )
    if not urls:
        return {}

    with ThreadPoolExecutor(max_workers=max_workers) as ex:
        raw_results = list(ex.map(_fetch_bytes, urls))

    cache = {}
    for url, raw in zip(urls, raw_results):
        cache[url] = load_image(url, raw_bytes=raw) if raw is not None else None
    return cache


# ==============================================================================
# TEXT UTILITIES
# ==============================================================================


def wrap_text_by_width(text, font_name, font_size, max_width):
    """
    Word-wraps text based on exact string pixel/point width rather than standard character count.
    """
    if not text:
        return [""]
    words, lines, current = text.split(), [], ""
    for word in words:
        candidate = f"{current} {word}".strip()
        if stringWidth(candidate, font_name, font_size) <= max_width:
            current = candidate
        else:
            if current:
                lines.append(current)
            current = word
    if current:
        lines.append(current)
    return lines or [""]


# ==============================================================================
# PDF DRAWING FUNCTIONS
# ==============================================================================


def draw_header(c, company_name, category_title):
    """
    Draws the main page masthead and column titles.
    Returns the y-coordinate where the first item row should start.
    """
    y = PAGE_H - MARGIN_TOP

    c.setFillColor(black)
    c.setFont("Helvetica-Bold", 15)
    c.drawString(MARGIN_L, y, company_name)
    y -= 0.35 * cm
    c.setLineWidth(1)
    c.line(MARGIN_L, y, PAGE_W - MARGIN_R, y)

    y -= 0.65 * cm
    c.setFont("Helvetica-Bold", 12.5)
    c.drawString(MARGIN_L, y, category_title)

    y -= 0.85 * cm
    if category_title != "Summary":
        c.setFont("Helvetica", 11)
        c.drawString(COL_NO_X, y, "No.")
        c.drawString(COL_CODE_X, y, "Code")
        c.drawString(COL_DESC_X, y, "Description")
        c.drawString(COL_QTY_X, y, "Quantity")
        y -= 0.25 * cm
        c.line(MARGIN_L, y, PAGE_W - MARGIN_R, y)

    return y - ROW_TOP_PAD


def draw_product_page(
    c,
    product,
    y,
    image_cache=None,
    company_name="BTM",
    category_title="KLS Martin Products",
):
    """
    Draws a single detailed product table row at the given y position.
    Automatically handles page breaking and header redrawing if space is insufficient.
    Returns the updated y-coordinate for the next row.
    """
    no = str(product.get("No.", ""))
    article = str(product.get("model", ""))
    qty = str(product.get("quantity", ""))
    description = product.get("description", "") or ""

    desc_lines = wrap_text_by_width(description, FONT_BODY, FONT_BODY_SIZE, DESC_COL_W)
    text_h = len(desc_lines) * LINE_HEIGHT

    img_url = product.get("image")
    if image_cache is not None and img_url in image_cache:
        img_reader = image_cache[img_url]
    else:
        img_reader = load_image(img_url)

    img_w = img_h = 0
    if img_reader is not None:
        iw, ih = img_reader.getSize()
        scale = min(IMG_MAX_W / iw, IMG_MAX_H / ih, 1.0)
        img_w, img_h = iw * scale, ih * scale

    content_h = max(text_h, img_h, MIN_ROW_CONTENT_H)
    row_block_h = content_h + ROW_BOTTOM_PAD

    if y - row_block_h < MARGIN_BOTTOM:
        c.showPage()
        y = draw_header(c, company_name, category_title)

    c.setFillColor(black)
    c.setFont(FONT_BODY, FONT_BODY_SIZE)
    first_baseline = y - FONT_BODY_SIZE
    c.drawString(COL_NO_X, first_baseline, no)
    c.drawString(COL_CODE_X, first_baseline, article)
    c.drawString(COL_QTY_X, first_baseline, qty)

    ty = first_baseline
    for line in desc_lines:
        c.drawString(COL_DESC_X, ty, line)
        ty -= LINE_HEIGHT

    if img_reader is not None:
        img_x = COL_IMG_X + (COL_IMG_W - img_w) / 2
        img_y = y - img_h
        c.drawImage(
            img_reader,
            img_x,
            img_y,
            width=img_w,
            height=img_h,
            preserveAspectRatio=True,
            mask="auto",
        )

    row_bottom = y - content_h
    line_y = row_bottom - ROW_BOTTOM_PAD
    c.setLineWidth(0.6)
    c.setStrokeColor(HexColor("#333333"))
    c.line(MARGIN_L, line_y, PAGE_W - MARGIN_R, line_y)

    return line_y - ROW_TOP_PAD


# --- Summary Table Component Helpers ---


def _draw_row_borders(c, col_bounds, top_y, bottom_y):
    """Draws grid lines around and between columns for table rows."""
    c.setLineWidth(TABLE_BORDER_WIDTH)
    c.setStrokeColor(TABLE_BORDER_COLOR)
    c.line(col_bounds[0], top_y, col_bounds[-1], top_y)
    c.line(col_bounds[0], bottom_y, col_bounds[-1], bottom_y)
    for x in col_bounds:
        c.line(x, top_y, x, bottom_y)


def draw_summary_table_header(
    c, y, col_bounds=SUMMARY_COL_BOUNDS, labels=SUMMARY_COL_LABELS
):
    """Draws the header row for the summary table."""
    top_y, bottom_y = y, y - TABLE_HEADER_H
    c.setFont("Helvetica-Bold", FONT_BODY_SIZE)
    c.setFillColor(black)
    baseline = bottom_y + (TABLE_HEADER_H - FONT_BODY_SIZE) / 2 + 2
    for i, label in enumerate(labels):
        c.drawString(col_bounds[i] + CELL_PAD_X, baseline, label)
    _draw_row_borders(c, col_bounds, top_y, bottom_y)
    return bottom_y


def draw_summary_row(
    c,
    product,
    y,
    company_name="BTM",
    category_title="Summary",
    col_bounds=SUMMARY_COL_BOUNDS,
):
    """
    Draws a single condensed row in the summary table view with automatic page break handling.
    """
    no = str(product.get("No.", ""))
    article = str(product.get("model", ""))
    qty = str(product.get("quantity", ""))
    description = product.get("description", "") or ""

    desc_lines = wrap_text_by_width(
        description, FONT_BODY, FONT_BODY_SIZE, SUMMARY_DESC_COL_W
    )
    content_h = max(len(desc_lines), 1) * LINE_HEIGHT
    row_h = content_h + 2 * CELL_PAD_Y

    if y - row_h < MARGIN_BOTTOM:
        c.showPage()
        y = draw_header(c, company_name, category_title)
        y = draw_summary_table_header(c, y, col_bounds)

    top_y, bottom_y = y, y - row_h
    c.setFillColor(black)
    c.setFont(FONT_BODY, FONT_BODY_SIZE)
    first_baseline = top_y - CELL_PAD_Y - FONT_BODY_SIZE

    c.drawString(col_bounds[0] + CELL_PAD_X, first_baseline, no)
    c.drawString(col_bounds[1] + CELL_PAD_X, first_baseline, article)
    c.drawString(col_bounds[3] + CELL_PAD_X, first_baseline, qty)

    ty = first_baseline
    for line in desc_lines:
        c.drawString(col_bounds[2] + CELL_PAD_X, ty, line)
        ty -= LINE_HEIGHT

    _draw_row_borders(c, col_bounds, top_y, bottom_y)
    return bottom_y

File: server/internal/service/pdf/pdf_drawer.go
// Package pdf renders Martin report data into PDF bytes. This package only
// builds PDF output from already-fetched data — no repository or HTTP calls
// beyond fetching the product images themselves, which are already-known
// URLs stored on each product.
//
// Ported from a reportlab (Python) script. ReportLab's canvas has y=0 at the
// page BOTTOM, increasing upward. fpdf has y=0 at the page TOP, increasing
// downward — every y-movement below was re-derived from actual up/down
// intent, not a mechanical sign flip. Two spots where a naive flip would be
// wrong: image anchor placement (ReportLab anchors bottom-left and grows up;
// fpdf anchors top-left and grows down — see drawProductRow), and the
// summary header's baseline centering (stays a subtraction in both).
package pdf

import (
	"bytes"
	"context"
	"fmt"
	"image"
	_ "embed"
	_ "image/jpeg" // registers jpeg decoding for image.DecodeConfig
	"io"
	"math"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/go-pdf/fpdf"

	"github.com/solenfw/calo-hub/internal/models"
)

// ==============================================================================
// CONFIGURATION & LAYOUT CONSTANTS
// ==============================================================================
//
// The document unit is points ("pt"), matching what ReportLab's canvas
// actually uses internally — `* cm` in the Python script was always just a
// convenience multiplier resolving to a point value at import time, same as
// here via the cm() helper. Bare numeric constants in the Python script
// (LINE_HEIGHT, TABLE_BORDER_WIDTH, font sizes) were already in points and
// are carried over unchanged.

const ptPerCm = 28.346456693

func cm(v float64) float64 { return v * ptPerCm }

var (
	pageW = cm(21.0) // A4
	pageH = cm(29.7)

	marginL      = cm(1.8)
	marginR      = cm(1.8)
	marginTop    = cm(1.6)
	marginBottom = cm(1.8)
	contentW     = pageW - marginL - marginR

	colNoX   = marginL
	colCodeX = marginL + cm(1.0)
	colDescX = marginL + cm(3.6)
	colQtyX  = marginL + cm(10.8)
	colImgX  = marginL + cm(12.6)
	colImgW  = contentW - cm(12.6)

	descColW = colQtyX - colDescX - cm(0.3)

	fontBody     = "Roboto"
	fontBodySize = 10.5
	lineHeight   = 14.0 // already points in the source script — unchanged

	imgMaxW = colImgW - cm(0.2)
	imgMaxH = cm(3.6)

	rowTopPad      = cm(0.55)
	rowBottomPad   = cm(0.55)
	minRowContentH = cm(2.3)

	tableHeaderH = cm(0.9)
	cellPadX     = cm(0.15)
	cellPadY     = cm(0.25)
	tableBorderW = 0.6 // bare in the source too — already points

	summaryColBounds = []float64{
		marginL,
		colCodeX,
		colDescX,
		colQtyX + cm(5),
		pageW - marginR,
	}
	summaryColLabels = []string{"No", "code", "Description", "Qty"}
	summaryDescColW  = (colQtyX + cm(5)) - colDescX - 2*cellPadX
)

const (
	maxImageWorkers   = 10
	imageFetchTimeout = 15 * time.Second
)

var httpClient = &http.Client{Timeout: imageFetchTimeout}

//go:embed assets/Roboto-Regular.ttf
var robotoRegularFontBytes []byte


// ==============================================================================
// ENTRY POINT
// ==============================================================================

// GeneratePDF renders a Martin report as a PDF: a detailed per-product page
// followed by a condensed summary table. Takes already-fetched product rows
// (via repository.GetMartinReportProductsByReportID) — no DB access happens
// in this package.
func GeneratePDF(ctx context.Context, companyName, reportName string, products []models.MartinReportProduct) ([]byte, error) {
	imageCache := prefetchImages(ctx, products)
	if len(robotoRegularFontBytes) == 0 {
		return nil, fmt.Errorf("pdf font bytes missing: embedded Roboto font not loaded")
	}

	doc := fpdf.New("P", "pt", "A4", "")
	doc.SetAutoPageBreak(false, 0) // pagination is handled manually below, same as the Python script's showPage() calls
	doc.AddUTF8FontFromBytes("Roboto", "", robotoRegularFontBytes)
	doc.AddUTF8FontFromBytes("Roboto", "B", robotoRegularFontBytes)
	doc.AddPage()
	y := drawHeader(doc, companyName, reportName)
	for _, p := range products {
		y = drawProductRow(doc, p, y, imageCache)
	}

	// add summary table.
	// doc.AddPage()
	// y = drawHeader(doc, companyName, "Summary")
	// y = drawSummaryTableHeader(doc, y)
	// for _, p := range products {
	// 	y = drawSummaryRow(doc, p, y, companyName)
	// }

	var buf bytes.Buffer
	if err := doc.Output(&buf); err != nil {
		return nil, fmt.Errorf("rendering pdf: %w", err)
	}
	return buf.Bytes(), nil
}

// ==============================================================================
// IMAGE FETCHING
// ==============================================================================

// prefetchImages downloads every distinct image URL concurrently (bounded
// worker pool, same pattern as cmd/migrate-images-local's download loop) and
// returns a url -> bytes cache. A failed fetch is cached as an explicit nil
// entry so drawProductRow treats it as "no image" instead of retrying.
func prefetchImages(ctx context.Context, products []models.MartinReportProduct) map[string][]byte {
	urlSet := make(map[string]struct{})
	for _, p := range products {
		if p.Image != nil && *p.Image != "" {
			urlSet[*p.Image] = struct{}{}
		}
	}
	if len(urlSet) == 0 {
		return nil
	}

	cache := make(map[string][]byte, len(urlSet))
	var mu sync.Mutex
	var wg sync.WaitGroup
	sem := make(chan struct{}, maxImageWorkers)

	for u := range urlSet {
		wg.Add(1)
		sem <- struct{}{}
		go func(url string) {
			defer wg.Done()
			defer func() { <-sem }()

			data, err := fetchImageBytes(ctx, url)
			mu.Lock()
			if err == nil {
				cache[url] = data
			} else {
				cache[url] = nil
			}
			mu.Unlock()
		}(u)
	}
	wg.Wait()
	return cache
}

func fetchImageBytes(ctx context.Context, url string) ([]byte, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("status %d", resp.StatusCode)
	}
	return io.ReadAll(resp.Body)
}

func imageDimensions(data []byte) (w, h float64, err error) {
	cfg, _, err := image.DecodeConfig(bytes.NewReader(data))
	if err != nil {
		return 0, 0, err
	}
	return float64(cfg.Width), float64(cfg.Height), nil
}

// ==============================================================================
// TEXT UTILITIES
// ==============================================================================

func wrapTextByWidth(doc *fpdf.Fpdf, text string, maxWidth float64) []string {
	if text == "" {
		return []string{""}
	}
	words := strings.Fields(text)
	var lines []string
	current := ""
	for _, word := range words {
		candidate := strings.TrimSpace(current + " " + word)
		if doc.GetStringWidth(candidate) <= maxWidth {
			current = candidate
		} else {
			if current != "" {
				lines = append(lines, current)
			}
			current = word
		}
	}
	if current != "" {
		lines = append(lines, current)
	}
	if len(lines) == 0 {
		lines = []string{""}
	}
	return lines
}

// ==============================================================================
// PDF DRAWING FUNCTIONS
// ==============================================================================

// drawHeader draws the page masthead and column titles, returning the y
// where the first row should start.
func drawHeader(doc *fpdf.Fpdf, companyName, categoryTitle string) float64 {
	y := marginTop

	doc.SetTextColor(0, 0, 0)
	doc.SetFont(fontBody, "B", 15)
	doc.Text(marginL, y, companyName)
	y += cm(0.35)
	doc.SetLineWidth(1)
	doc.Line(marginL, y, pageW-marginR, y)

	y += cm(0.65)
	doc.SetFont(fontBody, "B", 12.5)
	doc.Text(marginL, y, categoryTitle)

	y += cm(0.85)
	if categoryTitle != "Summary" {
		doc.SetFont(fontBody, "", 11)
		doc.Text(colNoX, y, "No.")
		doc.Text(colCodeX, y, "Code")
		doc.Text(colDescX, y, "Description")
		doc.Text(colQtyX, y, "Quantity")
		y += cm(0.25)
		doc.Line(marginL, y, pageW-marginR, y)
	}

	return y + rowTopPad
}

// drawProductRow draws one detailed product row at y, handling page breaks
// automatically. Returns the y for the next row.
func drawProductRow(doc *fpdf.Fpdf, product models.MartinReportProduct, y float64, imageCache map[string][]byte) float64 {
	doc.SetFont(fontBody, "", fontBodySize)
	descLines := wrapTextByWidth(doc, product.Description, descColW)
	textH := float64(len(descLines)) * lineHeight

	var imgData []byte
	if product.Image != nil {
		imgData = imageCache[*product.Image]
	}

	haveImage := imgData != nil
	var imgW, imgH float64
	if haveImage {
		iw, ih, err := imageDimensions(imgData)
		if err != nil || iw <= 0 || ih <= 0 {
			haveImage = false
		} else {
			// Same scaling approach as the Python script: pixel dimensions
			// are treated as point-equivalent for the fit calculation,
			// preserved as-is rather than made DPI-aware.
			scale := math.Min(imgMaxW/iw, imgMaxH/ih)
			if scale > 1.0 {
				scale = 1.0
			}
			imgW, imgH = iw*scale, ih*scale
		}
	}

	contentH := math.Max(textH, math.Max(imgH, minRowContentH))
	rowBlockH := contentH + rowBottomPad

	if y+rowBlockH > pageH-marginBottom {
		doc.AddPage()
		y = drawHeader(doc, "", "")
	}

	doc.SetTextColor(0, 0, 0)
	doc.SetFont(fontBody, "", fontBodySize)
	firstBaseline := y + fontBodySize
	doc.Text(colNoX, firstBaseline, fmt.Sprintf("%d", product.RowNo))
	doc.Text(colCodeX, firstBaseline, product.Code)
	doc.Text(colQtyX, firstBaseline, fmt.Sprintf("%d", product.Quantity))

	ty := firstBaseline
	for _, line := range descLines {
		doc.Text(colDescX, ty, line)
		ty += lineHeight
	}

	if haveImage {
		imgX := colImgX + (colImgW-imgW)/2
		// fpdf anchors images top-left and grows downward, so the row's
		// top (y) IS the image's top edge directly — no subtraction, unlike
		// the Python version's `y - img_h` (which computed a bottom-left
		// anchor for ReportLab's upward-growing images).
		imgY := y
		placeImage(doc, fmt.Sprintf("row-%d", product.RowNo), imgData, imgX, imgY, imgW, imgH)
	}

	rowBottom := y + contentH
	lineY := rowBottom + rowBottomPad
	doc.SetLineWidth(tableBorderW)
	doc.SetDrawColor(0x33, 0x33, 0x33)
	doc.Line(marginL, lineY, pageW-marginR, lineY)

	return lineY + rowTopPad
}

// placeImage registers image bytes under a name unique to this document
// build (RowNo is unique within one report's product slice, and a fresh
// *fpdf.Fpdf is created per GeneratePDF call, so no collision risk).
func placeImage(doc *fpdf.Fpdf, name string, data []byte, x, y, w, h float64) {
	opt := fpdf.ImageOptions{ImageType: "JPG"}
	doc.RegisterImageOptionsReader(name, opt, bytes.NewReader(data))
	doc.ImageOptions(name, x, y, w, h, false, opt, 0, "")
}

// --- Summary table ---

func drawRowBorders(doc *fpdf.Fpdf, colBounds []float64, topY, bottomY float64) {
	doc.SetLineWidth(tableBorderW)
	doc.SetDrawColor(0x33, 0x33, 0x33)
	doc.Line(colBounds[0], topY, colBounds[len(colBounds)-1], topY)
	doc.Line(colBounds[0], bottomY, colBounds[len(colBounds)-1], bottomY)
	for _, x := range colBounds {
		doc.Line(x, topY, x, bottomY)
	}
}

func drawSummaryTableHeader(doc *fpdf.Fpdf, y float64) float64 {
	topY, bottomY := y, y+tableHeaderH
	doc.SetFont(fontBody, "B", fontBodySize)
	doc.SetTextColor(0, 0, 0)
	// Centering offset stays a SUBTRACTION here — this is the other spot
	// where a mechanical sign-flip would be wrong. See the package comment.
	baseline := bottomY - (tableHeaderH-fontBodySize)/2 - 2
	for i, label := range summaryColLabels {
		doc.Text(summaryColBounds[i]+cellPadX, baseline, label)
	}
	drawRowBorders(doc, summaryColBounds, topY, bottomY)
	return bottomY
}

func drawSummaryRow(doc *fpdf.Fpdf, product models.MartinReportProduct, y float64, companyName string) float64 {
	doc.SetFont(fontBody, "", fontBodySize)
	descLines := wrapTextByWidth(doc, product.Description, summaryDescColW)
	contentH := math.Max(float64(len(descLines)), 1) * lineHeight
	rowH := contentH + 2*cellPadY

	if y+rowH > pageH-marginBottom {
		doc.AddPage()
		y = drawHeader(doc, companyName, "")
		y = drawSummaryTableHeader(doc, y)
	}

	topY, bottomY := y, y+rowH
	doc.SetTextColor(0, 0, 0)
	doc.SetFont(fontBody, "", fontBodySize)
	firstBaseline := topY + cellPadY + fontBodySize

	doc.Text(summaryColBounds[0]+cellPadX, firstBaseline, fmt.Sprintf("%d", product.RowNo))
	doc.Text(summaryColBounds[1]+cellPadX, firstBaseline, product.Code)
	doc.Text(summaryColBounds[3]+cellPadX, firstBaseline, fmt.Sprintf("%d", product.Quantity))

	ty := firstBaseline
	for _, line := range descLines {
		doc.Text(summaryColBounds[2]+cellPadX, ty, line)
		ty += lineHeight
	}

	drawRowBorders(doc, summaryColBounds, topY, bottomY)
	return bottomY
}